checker/src/org/checkerframework/checker/guieffect/GuiEffectVisitor.java

```diff
@@ -158,16 +227,75 @@ public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
             // Instantiate type-polymorphic effects
             if (srcType.hasAnnotation(AlwaysSafe.class)) {
                 targetEffect = new Effect(SafeEffect.class);
+                polyUIArgumentsAreUI = false;
             } else if (srcType.hasAnnotation(UI.class)) {
                 targetEffect = new Effect(UIEffect.class);
             }
             // Poly substitution would be a noop.
         }
 
-        Effect callerEffect = atypeFactory.getDeclaredEffect(callerElt);
-        // Field initializers inside anonymous inner classes show up with a null current-method ---
-        // the traversal goes straight from the class to the initializer.
-        assert (currentMethods.peek() == null || callerEffect.equals(effStack.peek()));
+        // Look for lambda arguments and infer them to be @UIEffect for this call
+        List<? extends ExpressionTree> args = node.getArguments();
+        List<? extends VariableElement> parameters = methodElt.getParameters();
+        Pair<AnnotatedExecutableType, List<AnnotatedTypeMirror>> mfuPair =
+                atypeFactory.methodFromUse(node);
+        AnnotatedExecutableType invokedMethod = mfuPair.first;
+        List<AnnotatedTypeMirror> typeargs =
+                AnnotatedTypes.expandVarArgs(atypeFactory, invokedMethod, node.getArguments());
+        for (int i = 0; i < args.size(); ++i) {
+            if (args.get(i) instanceof LambdaExpressionTree) {
+                //System.err.println(methodElt + " is from stub file? " + atypeFactory.isFromStubFile(methodElt));
+                LambdaExpressionTree lambdaTree = (LambdaExpressionTree) args.get(i);
+                VariableElement formalArgElement = parameters.get(i);
+                // For some reason, these two have different information, and neither is a subset of the other.
+                // argTypeFromMFU is needed for stubs, argTypeFromArgElement will actually find @PolyUI when in the
+                // source. Neither will find @PolyUI in stubs.
+                // ToDo: Figure out a better way to works in both cases
+                AnnotatedTypeMirror argTypeFromArgElement =
+                        atypeFactory.getAnnotatedType(formalArgElement);
+                AnnotatedTypeMirror argTypeFromMFU = typeargs.get(i);
+                ExecutableElement argFIMethodElt =
+                        TreeUtils.getFunctionalInterfaceMethod(lambdaTree);
+                Effect lambdaEffect = atypeFactory.getDeclaredEffect(argFIMethodElt);
+                // How are these not a at least one a super set of the other?
+                /*System.err.println(methodElt
+                + " UI_from_mfu=" + argTypeFromMFU.hasAnnotation(UI.class)
+                + " PolyUI_from_mfu=" + argTypeFromMFU.hasAnnotation(PolyUI.class)
+                + " UI_from_argElem=" + argTypeFromArgElement.hasAnnotation(UI.class)
+                + " PolyUI_from_argElem=" + argTypeFromArgElement.hasAnnotation(PolyUI.class));*/
```

Comment:
This top-down inference approach will always be a bit fragile (regarding your comment about many nested applications of the polymorphic identity function).  Some depth of nested arguments will remain unannotated when you first visit the outer lambda applications, leading to calling some lambda applications safe when they should be UI.  (This is sound, just a little brittle and less expressive than it might be.)

Did you consider moving this after the recursive visitation, so all of the nested subexpressions will already have inference complete?  Then the contexts would then have incomplete information instead of the nested parameters, but I'm wondering if this was a choice or the first thing that you tried that worked well on your programs of interest.

Reply:
I am happy to switch to bottom-up. I think it makes a lot of sense. Will rework the code to do that, but see above regarding `GuiEffectTreeAnnotator` and annotating lambda expressions themselves.